tests/SAML2/XML/samlp/RequestedAuthnContextTest.php

```diff
@@ -0,0 +1,102 @@
+        $requestedAuthnContext->setComparison('exact');
+        $requestedAuthnContext->setRequestedAuthnContexts([$authnContextDeclRef]);
+
+        $document = DOMDocumentFactory::fromString('<root />');
+        /** @psalm-var \DOMElement $document->firstChild */
+        $requestedAuthnContextElement = $requestedAuthnContext->toXML($document->firstChild);
+
+        $this->assertEquals('exact', $requestedAuthnContextElement->getAttribute('Comparison'));
+
+        $authnContextDeclRefElements = Utils::xpQuery(
+            $requestedAuthnContextElement,
+            './saml_assertion:AuthnContextDeclRef'
+        );
+
+        $this->assertCount(1, $authnContextDeclRefElements);
+        $authnContextDeclRefElement = $authnContextDeclRefElements[0];
+
+        $this->assertEquals('/relative/path/to/document.xml', $authnContextDeclRefElement->textContent);
+    }
+
+
+    /**
+     * @return void
+     */
+    public function testUnmarshalling(): void
+    {
+        $samlNamespace = Constants::NS_SAML;
+        $samlpNamespace = Constants::NS_SAMLP;
+        $document = DOMDocumentFactory::fromString(<<<XML
+<samlp:RequestedAuthnContext xmlns:samlp="{$samlpNamespace}" Comparison="minimum">
+  <saml:AuthnContextClassRef xmlns:saml="{$samlNamespace}">
+    urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport
+  </saml:AuthnContextClassRef>
+</samlp:RequestedAuthnContext>
+XML
+        );
+
+        /** @psalm-var \DOMElement $document->firstChild */
+        $requestedAuthnContext = new RequestedAuthnContext($document->firstChild);
+        $this->assertEquals('minimum', $requestedAuthnContext->getComparison());
+
+        $contexts = $requestedAuthnContext->getRequestedAuthnContexts();
+        $this->assertCount(1, $contexts);
+        $this->assertInstanceOf(AuthnContextClassRef::class, $contexts[0]);
+        $this->assertEquals(Constants::AC_PASSWORD_PROTECTED_TRANSPORT, $contexts[0]->getClassRef());
+    }
+
+
+    /**
+     * @return void
+     */
+    public function testUnmarshallingWithMixedContextsFails(): void
+    {
+        $samlNamespace = Constants::NS_SAML;
+        $samlpNamespace = Constants::NS_SAMLP;
+
+        $document = DOMDocumentFactory::fromString(<<<XML
+<samlp:RequestedAuthnContext xmlns:samlp="{$samlpNamespace}" Comparison="minimum">
+  <saml:AuthnContextClassRef xmlns:saml="{$samlNamespace}">
+    urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport
+  </saml:AuthnContextClassRef>
+  <saml:AuthnContextDeclRef xmlns:saml="{$samlNamespace}">
+    /relative/path/to/document.xml
+  </saml:AuthnContextDeclRef>
+</samlp:RequestedAuthnContext>
+XML
+        );
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('You need either AuthnContextClassRef or AuthnContextDeclRef, not both');
+
+        /** @psalm-var \DOMElement $document->firstChild */
+        $requestedAuthnContext = new RequestedAuthnContext($document->firstChild);
+    }
```

Comment:
We should probably have a test for the opposite scenario, trying to marshall into XML when you have added both `AuthnContextClassRef` and `AuthnContextDeclrRef`.

Reply:
Consider it done!